### PR TITLE
feat: split downloads and verification into two steps

### PR DESCRIPTION
**Pull request by vmx**
This is a clear separation of concerns. First get the prerequisites
for the verification, then do the actual verification.

The newly introduced download script is very similar to the
verification script, you can diff between them for easier review
of the code.

I haven't had a test run with it yet, but I'd like to get a review ASAP
though (also on the general idea).

### CODE REVIEW

**Comment by vmx** [comment]
I verified it locally, it works for me.

**Comment by vmx** [comment]
This PR is now in a state where it could be merged. It should work (also the downloads) as it currently is.

File: b288702/download_prereqs_contrib.sh [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    fi
+
+    if [[ ! -f ${phase1_file} ]]; then
+        log "downloading Phase 1 parameters: ${phase1_file}"
+        curl --progress-bar -O "${url_base_phase1}/${phase1_file}"
+    fi
+    log 'verifying Phase 1 checksum'
+    echo "${phase1_checksum} ${phase1_file}" | b2sum -c
+
+    log "${green}success:${off} finished generating initial phase2 parameters"
```
Comment (porcuquine) [inline b288702/download_prereqs_contrib.sh:104]: This reports that the intial phase2 parameters have been generated, but the code doing that seems to have been removed. It previously happened in the (now completely removed) `verify_intial.sh` script: https://github.com/filecoin-project/phase2-attestations/blob/master/b288702/verify_initial.sh#L82-L98

It's not clear to me that that script should have been removed, but if we *do* remove it, we certainly need to replace the code to generate the initial parameters. The reason I don't think that code belongs here is that it's very explicitly *not* a download (which is what #23) was ensuring.

Comment (porcuquine) [reply b288702/download_prereqs_contrib.sh:104]: [Having read further, I see that the code is moved, and this is a stray log.]

Comment (vmx) [reply b288702/download_prereqs_contrib.sh:104]: This log line needs to be removed. The parameter generation is now happen in the `verify_contribs.sh` script.

The reason to move it into `verify_contribs.sh` is, that the `verify_initial.sh` script didn't distinguish between contributions. It created the parameters, even if you would verify only a single distinct contribution.

Comment (porcuquine) [reply b288702/download_prereqs_contrib.sh]: Understood. The `verify_initial.sh` was only meant to be run once per circuit — and that is what the instructions described. If this were to be moved into `verify_contrib.sh` we would either generate the initial params every time (very wasteful) or else have to have detection there.

What we actually want is to generate the initial parameters exactly once per circuit. It's logically equivalent to downloading (in terms of providing a run-once method to the `0` params) but must replace it.

File: b288702/verify_contrib.sh [hunk 2]
```diff
@@ -81,35 +75,72 @@ if [[ ! -f './b288702.b2sums' ]]; then
...
+    if [[ ! -f ${prev_file} ]]; then
+        error "${prev_file} is missing. Run: ./download_prereqs_contrib.sh ${proof} ${sector_size} ${contrib}"
+        exit 1
+    fi
+    # Verify checksum even if file was present, in case of incomplete download
+    # or corruption. This is especially relevant if another process might have
+    # initiated a download now in process.
+    log 'verifying downloaded params checksum'
+    grep $prev_file b288702.b2sums | b2sum -c
+fi
```
Comment (porcuquine) [inline b288702/verify_contrib.sh:131]: As noted above, this should be its own script the verifier runs exactly once. That script can probably have  a better name the `verify_initial.sh`, though. I think `generate_initial.sh` is probably good.

Comment (vmx) [reply b288702/verify_contrib.sh:131]: The of having it there is actually that you don't necessarily want to run it. If you split the work and you only want to verify contribution number 5 of a circuit, it would be a waste to create those params.

You could of course not run the `generate_initial.sh` script. But it would make the description way more difficult then just running the generation conditionally based on the contribution you want to verify.

It can well be that I'm missing something and there is a reason to generate them, even if you do not verify the first contribution.

Comment (porcuquine) [reply b288702/verify_contrib.sh:131]: I understand the point you're making, and I can probably be convinced. Whichever way we go, *something* must be documented or changed in the README.

Either:
- We make initial parameter-generation explicit, and instruct the verifier to perform it only before verifying the first contribution — something like:
```
./generate_initial.sh && ./verify_contrib.sh winning 32 1
```
Or:
- We add documentation to the effect that verifying contribution 1 is more expensive and is actually performing two distinct steps.

I think it's cleaner and clearer to document the special case by changing the commands as in the first option above. This makes clear the actual process so the verifier understands what is happening. If the verifier *does* understand the process, they should expect to see the initial parameter generation somewhere, and putting the command at top level will make it easier for them to confirm that expectation by inspection.

Comment (ribasushi) [reply b288702/verify_contrib.sh:131]: Oh that's easy: I *very much* prefer the first version. It shows the *logical* discrete steps, not necessarily the nitty-gritty, but it gives you an idea of "oh... this thing is quite different".

It's not only about "how long it takes", it is also about "what logs scroll by"

Remember that this reorg started *precisely* because I paid attention to the logs and went "wait... this doesn't... at all agree with what I think I am doing here". Transparency/verbosity in the context of verification is **really** good, as long as it does not become overwhelming.

TLDR: 
`./generate_initial.sh winning 32 && ./verify_contrib.sh winning 32 1`
is **vastly** preferable to
`./verify-stuff-trust-me-it-is-all-automated-mkay winning 32`

I mean heck - I might *want* to disconnect the network cable between the steps, who knows what makes me (the verifier) happy.

Comment (porcuquine) [reply b288702/verify_contrib.sh:131]: Thanks, @ribasushi. My opinions was, of course, informed by our discussion of these considerations. Since incorporating your feedback in order to ensure your verification pass provides the assurance we want (of both correctness *and* of confidence in the process), I consider your preference controlling. Let's go with the first option.

Comment (porcuquine) [reply b288702/verify_contrib.sh:131]: A slight amendment: we also need to include the download of phase1 parameters (which my example above did not). The existing overloaded download script (with special case for contribution 1) could be used, or we could be even more explicit and break that out into its own script. This leads to either:

```
./download_prereqs_generate_initial.sh winning 32 && ./generate_initial.sh winning 32 && ./verify_contrib.sh winning 32 1
```
Or:
```
./download_prereqs_contrib.sh winning 32 1 && ./generate_initial.sh winning 32 && ./verify_contrib.sh winning 32 1
```

Again, I favor the first option.

Comment (porcuquine) [reply b288702/verify_contrib.sh:131]: Per discussion with @ribasushi I am amending the recommended first option to avoid grammatical ambiguity:
```
./download_initial_generation_prereqs.sh winning 32 && ./generate_initial.sh winning 32 && ./verify_contrib.sh winning 32 1
```

Comment (vmx) [reply b288702/verify_contrib.sh:131]: To make sure we are on the same page. I understand and agree. The other contributions (other then the first one) will still be:

    ./download_prereqs_contrib.sh winning 32 7 && ./verify_contrib.sh winning 32 7

correct?

Comment (porcuquine) [reply b288702/verify_contrib.sh:131]: Yes — only the first contribution will change.

Comment (porcuquine) [reply b288702/verify_contrib.sh:131]: (For each circuit.)

**Comment by vmx** [comment]
@porcuquine @ribasushi here's the new version. I haven't tried it locally yet, but I'll do now. Though you can already have a look if that's what you both were looking for.

**Comment by vmx** [comment]
I tried it locally. From my POV it would be ready to go.

File: b288702/README.md [hunk 3]
```diff
@@ -93,132 +95,123 @@ other. Because each verification step requires two sequential contribution files
...
+  ./download_prereqs_contrib.sh winning 32 11 && ./verify_contrib.sh winning 32 11
+  ./download_prereqs_contrib.sh winning 32 12 && ./verify_contrib.sh winning 32 12
+  ./download_prereqs_contrib.sh winning 32 13 && ./verify_contrib.sh winning 32 13
+  ./download_prereqs_contrib.sh winning 32 14 && ./verify_contrib.sh winning 32 14
+  ./download_prereqs_contrib.sh winning 32 15 && ./verify_contrib.sh winning 32 15
+  ./download_prereqs_contrib.sh winning 32 16 && ./verify_contrib.sh winning 32 16
+  ./download_prereqs_contrib.sh winning 32 17 && ./verify_contrib.sh winning 32 17
+  ./download_prereqs_contrib.sh winning 32 18 && ./verify_contrib.sh winning 32 18
+  ./download_prereqs_contrib.sh winning 32 19 && ./verify_contrib.sh winning 32 19
+  ./download_prereqs_contrib.sh winning 32 20 && ./verify_contrib.sh winning 32 20
```
Comment (ribasushi) [inline b288702/README.md:121]: ( could not resist )

Charlie-Sheen-Winning-Poster

File: b288702/download_initial_generation_prereqs.sh [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+fi
+log 'verifying Phase 1 checksum'
+echo "${phase1_checksum} ${phase1_file}" | b2sum -c
+
+# Download the parameters of the first contribution
+file="${proof}_poseidon_${sector_size}gib_b288702_1_small_raw"
+if [[ ! -f ${file} ]]; then
+    log "downloading params: ${file}"
+    curl --progress-bar -O "${url_base}/${file}"
+fi
```
Comment (porcuquine) [inline b288702/download_initial_generation_prereqs.sh:87]: I guess for greatest consistency, this would still be in `download_prereqs_contrib.sh`. I think I can live with it here, though. @ribasushi ?

**Review by ribasushi — APPROVED** [review]
I think it looks right, have not tried to run things yet though, let's merge it first.

File: b288702/download_initial_generation_prereqs.sh (hunk 4 above)
Comment (ribasushi) [reply b288702/download_initial_generation_prereqs.sh:87]: @porcuquine actually I requested it to be separate. Needing to download it on *every* step is wasteful ( the matrix is *huge* ). Since it is needed on step 1 only - I asked @vmx to download it only when I will be generating "0_large"

Comment (vmx) [reply b288702/download_initial_generation_prereqs.sh:87]: @porcuquine Do you want to have a `./download_prereqs_contrib.sh` call also in the inital step? Like this:

./download_initial_generation_prereqs.sh winning 32 && ./generate_initial.sh winning 32 && **./download_prereqs_contrib.sh winning 32 1** && ./verify_contrib.sh winning 32 1

Instead of what we are currently doing:

./download_initial_generation_prereqs.sh winning 32 && ./generate_initial.sh winning 32 && ./verify_contrib.sh winning 32 1

Comment (ribasushi) [reply b288702/download_initial_generation_prereqs.sh:87]: @vmx the more distinct steps there are in the ` ... && ...` list - the better
There already is one "magic script" of ./verify-all. Further obscuring at intermediate steps is not reasonable.

Comment (vmx) [reply b288702/download_initial_generation_prereqs.sh:87]: > @vmx the more distinct steps there are in the ` ... && ...` list - the better

Which means? Adding the additional `./download_prereqs_contrib.sh winning 32 1` would mean that this script would get a special case.

Comment (ribasushi) [reply b288702/download_initial_generation_prereqs.sh:87]: I think this is ok / desired.
@porcuquine thoughts?

Comment (porcuquine) [reply b288702/download_initial_generation_prereqs.sh:87]: @ribasushi 
> Needing to download it on every step is wasteful ( the matrix is huge ). Since it is needed on step 1 only - I asked @vmx to download it only when I will be generating "0_large"

I am not sure I follow. I think the confusion might be that the lines I intended to comment on are the ones which download the step 1 file:
`file="${proof}_poseidon_${sector_size}gib_b288702_1_small_raw"`.

NOTE: I am catching up to this conversation and will try to answer things in order, so my first responses may be outdated.

Comment (porcuquine) [reply b288702/download_initial_generation_prereqs.sh:87]: > Do you want to have a ./download_prereqs_contrib.sh call also in the inital step? Like this:
>
> ```./download_initial_generation_prereqs.sh winning 32 && ./generate_initial.sh winning 32 && ./download_prereqs_contrib.sh winning 32 1 && ./verify_contrib.sh winning 32 1```

Yes, that is what I was suggesting. I think this is the organization which was logically implied by all the discussions we had yesterday. We just didn't quite fully break it down on the first pass, and I didn't notice until I made that review comment.

Comment (porcuquine) [reply b288702/download_initial_generation_prereqs.sh:87]: > Which means? Adding the additional ./download_prereqs_contrib.sh winning 32 1 would mean that this script would get a special case.

> I think this is ok / desired.

Yes, agreed. For clarity, the special case is that we should never download the 'before' (0) params when verifying the first (1) contribution. We should always use the generated ones. We can still verify the checksum as usual and fail if the generated params were not as expected.

Comment (ribasushi) [reply b288702/download_initial_generation_prereqs.sh:87]: > We should always use the generated ones.

Not sure if this is the right time/place: but perhaps ./phase2 shouldn't be so "magical/helpful" and instead require all its arguments to be explicitly specified.

Not a request, just thinking out-loud

Comment (vmx) [reply b288702/download_initial_generation_prereqs.sh:87]: I think I got the idea, I'll do a PR.
